snuba/datasets/all_events.py

```diff
@@ -0,0 +1,165 @@
+from snuba.query.query_processor import QueryProcessor
+from snuba.query.timeseries import TimeSeriesExtension
+from snuba.request.request_settings import RequestSettings
+from snuba.util import columns_in_expr, is_condition
+
+
+EVENTS = 'events'
+TRANSACTIONS = 'transactions'
+TRANSACTIONS_ONLY_COLUMNS = [
+    'trace_id',
+    'span_id',
+    'transaction_name',
+    'transaction_hash',
+    'transaction_op',
+    'start_ts',
+    'start_ms',
+    'finish__ts',
+    'finish_ms',
+    'duration',
+]
+
+EVENTS_ONLY_COLUMNS = [
+    'group_id',
+    'transaction',
+]
+
+
+# Based on query.get_all_referenced_columns()
+def get_condition_columns(query: Query) -> Sequence[Any]:
+    col_exprs: MutableSequence[Any] = []
+    # Conditions need flattening as they can be nested as AND/OR
+    if query.get_conditions():
+        flat_conditions = list(
+            chain(*[[c] if is_condition(c) else c for c in query.get_conditions()]))
+        col_exprs.extend([c[0] for c in flat_conditions])
+
+    # Return the set of all columns referenced in any expression
+    return set(chain(*[columns_in_expr(ex) for ex in col_exprs]))
+
+
+def detect_dataset(query: Query) -> str:
+    """
+    Given a query, we attempt to guess whether it is better to fetch data from the
+    "events" or "transactions" dataset. This is going to be wrong in some cases.
+    """
+    # First check for a top level condition that matches either type = transaction
+    # type != transaction.
+    conditions = query.get_conditions()
+    if conditions:
+        for idx, condition in enumerate(conditions):
+            if is_condition(condition):
+                if tuple(condition) == ('type', '!=', 'transaction'):
+                    return EVENTS
+                elif tuple(condition) == ('type', '=', 'transaction'):
+                    return TRANSACTIONS
+
+    # If there is a condition that references a transactions only field, just switch
+    # to the transactions dataset
+    if [col for col in TRANSACTIONS_ONLY_COLUMNS if col in get_condition_columns(query)]:
+        return TRANSACTIONS
+
+    # Use events by default
+    return EVENTS
+
+
+class AllEventsProcessor(QueryProcessor):
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        """
+        Switches the data source from the default (Events) to the transactions
+        table if a transaction specific column is detected.
+        Sets all the other columns to none
+
+        """
+        detected_dataset = detect_dataset(query)
+
+        source = get_dataset(detected_dataset) \
+            .get_dataset_schemas() \
+            .get_read_schema() \
+            .get_data_source()
+        query.set_data_source(source)
+
+        if detected_dataset == TRANSACTIONS:
+            with_columns = [
+                ("'transaction'", 'type'),
+                ('finish_ts', 'timestamp'),
+                ('user_name', 'username'),
+                ('user_email', 'email'),
```

Comment:
what's the advantage of using a with instead of just replacing the columns in the query object itself (besides the fact that I did not yet provide a conveniente `replace_column_in_query` method to use and it has to be done manually)? 
Going on, replacing the column names is going to be a more reliable solution since it leaves a query that is easier to process for the next processor. Using the WITH clause will require the next processor to understand the WITH and figure out what changes to do.

Reply:
We can use `column_expr` to achieve this today even before we get the new method. However there are some edge cases that WITH solves that I need to verify for the column replacement solution. OTOH these will be things like-
- columns like `uniq(colname)`
- column not referenced anywhere except in orderby
- having clause

Reply:
I'd advise against using column_expr is not strictly necessary since it works at a different point in the query processing (during formatting basically), so it will be harder to move that back.
About the column above #555 should solve them. If you need a stop gap solution, ok, let's add them to column_expr until the PR is committed, but let's ensure 555 works for you so we can move them back easily.

About the HAVING clause. I think we can support it in #555. My theory is that it was just forgotten from all_referenced_columns. It is a quick fix

'column not referenced anywhere except in orderby' what's the problem here?

`uniq(colname)` I think you will still see such a column in all_referenced_column and can replace it in #555. Do you have a specific example where this does not work ?

Reply:
Sorry, it was just a note to self of some cases I wanted to remember to test against the column_expr implementation. Seems to all be working fine.